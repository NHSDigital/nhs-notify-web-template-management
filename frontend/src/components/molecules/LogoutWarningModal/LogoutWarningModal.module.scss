.signIn {
  width: 100% !important; // This is because nhsuk styles overwrite width.
}

.signOut {
  font-size: 1.1em;
  font-weight: 600;
  text-align: center;
}
