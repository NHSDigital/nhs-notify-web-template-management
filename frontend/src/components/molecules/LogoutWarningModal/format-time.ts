'use client';

export const formatTime = (remainingSeconds: number) => {
  if (remainingSeconds <= 30) {
    return `${remainingSeconds} seconds`;
  }

  if (remainingSeconds <= 60) {
    return '1 minute';
  }

  return `${Math.ceil(remainingSeconds / 60)} minutes`;
};
