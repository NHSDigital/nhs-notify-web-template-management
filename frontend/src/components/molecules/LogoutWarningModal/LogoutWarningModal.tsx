'use client';

import { useEffect, useRef, useState } from 'react';
import { useIdleTimer } from 'react-idle-timer';
import { Button } from 'nhsuk-react-components';
import { fetchAuthSession } from 'aws-amplify/auth';
import { useAuthenticator } from '@aws-amplify/ui-react';
import { useRouter, usePathname } from 'next/navigation';
import { Modal } from '@molecules/Modal/Modal';
import content from '@content/content';
import { getBasePath } from '@utils/get-base-path';
import { formatTime } from './format-time';

import styles from './LogoutWarningModal.module.scss';

export const LogoutWarningModal = ({
  promptBeforeLogoutSeconds: promptTimeSeconds,
  logoutInSeconds,
}: {
  promptBeforeLogoutSeconds: number;
  logoutInSeconds: number;
}) => {
  const {
    headerComponent: {
      links: { logOut },
    },
    logoutWarningComponent,
  } = content.components;

  const router = useRouter();
  const pathname = usePathname();
  const initial = formatTime(promptTimeSeconds);
  const [showModal, setShowModal] = useState(false);
  const [remainingTime, setRemainingTime] = useState(initial);
  const { authStatus } = useAuthenticator((ctx) => [ctx.authStatus]);

  const idle = () => {
    setShowModal(false);

    router.push(
      `/auth/inactive?redirect=${encodeURIComponent(
        `${getBasePath()}/${pathname}`
      )}`
    );
  };

  const prompted = () => {
    setShowModal(true);
  };

  const { reset, getRemainingTime } = useIdleTimer({
    timeout: logoutInSeconds * 1000,
    promptBeforeIdle: promptTimeSeconds * 1000,
    onPrompt: prompted,
    onIdle: idle,
    disabled: authStatus !== 'authenticated',
  });

  const intervalRef = useRef<ReturnType<typeof setInterval>>();

  useEffect(() => {
    if (!showModal) return;

    intervalRef.current = setInterval(() => {
      const remaining = Math.ceil(getRemainingTime() / 1000);

      setRemainingTime(formatTime(remaining));
    }, 1000);

    return () => {
      clearInterval(intervalRef.current);
    };
  }, [showModal, getRemainingTime]);

  const stillHere = () => {
    reset();
    setShowModal(false);
    setRemainingTime(initial);
    fetchAuthSession({ forceRefresh: true });
  };

  return (
    <Modal showModal={showModal}>
      <Modal.Header>
        {`${logoutWarningComponent.heading} ${remainingTime}.`}
      </Modal.Header>
      <Modal.Body>
        <p>{logoutWarningComponent.body}</p>
      </Modal.Body>
      <Modal.Footer>
        <Button
          className={styles.signIn}
          onClick={stillHere}
          data-testid='modal-sign-in'
        >
          {logoutWarningComponent.signIn}
        </Button>
        <div className={styles.signOut}>
          <a
            className='nhsuk-link'
            href={logOut.href}
            data-testid='modal-sign-out'
          >
            {logOut.text}
          </a>
        </div>
      </Modal.Footer>
    </Modal>
  );
};
