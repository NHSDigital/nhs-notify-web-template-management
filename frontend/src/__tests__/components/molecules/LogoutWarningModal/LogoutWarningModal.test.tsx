import { render, fireEvent, act } from '@testing-library/react';
import { LogoutWarningModal } from '@molecules/LogoutWarningModal/LogoutWarningModal';
import { useAuthenticator, type UseAuthenticator } from '@aws-amplify/ui-react';
import { useRouter, usePathname } from 'next/navigation';
import { mockDeep } from 'jest-mock-extended';

jest.mock('next/navigation');
jest.mock('aws-amplify/auth');
jest.mock('@aws-amplify/ui-react');

const useAuthenticatorMock = jest.mocked(useAuthenticator);

const routerPushMock: jest.Mock = jest.fn();

const dialogCloseMock = jest.fn(function mock(this: HTMLDialogElement) {
  this.open = false;
});

const dialogOpenMock = jest.fn(function mock(this: HTMLDialogElement) {
  this.open = true;
});

describe('LogoutWarningModal', () => {
  beforeAll(() => {
    jest.resetAllMocks();

    // Polyfill dialog actions since JSDOM does not support it
    // https://github.com/jsdom/jsdom/issues/3294
    HTMLDialogElement.prototype.showModal = dialogOpenMock;

    HTMLDialogElement.prototype.close = dialogCloseMock;

    useAuthenticatorMock.mockImplementation((cb) => {
      const context = mockDeep<UseAuthenticator>({
        authStatus: 'authenticated',
      });

      if (cb) {
        cb(context);
      }

      return context;
    });

    (useRouter as jest.Mock).mockReturnValue({
      push: routerPushMock,
    });

    (usePathname as jest.Mock).mockReturnValue('testing');

    jest.useFakeTimers();
  });

  test('should match snapshot', async () => {
    const container = render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  test('should stay closed if not authenticated', () => {
    useAuthenticatorMock.mockReturnValueOnce(
      mockDeep<UseAuthenticator>({
        authStatus: 'unauthenticated',
      })
    );

    render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    act(() => {
      // advance timers by 15 minutes
      jest.advanceTimersByTime(60 * 15 * 1000);
      fireEvent.focus(document);
    });

    expect(routerPushMock).not.toHaveBeenCalled();

    expect(dialogOpenMock).not.toHaveBeenCalled();

    expect(dialogCloseMock).toHaveBeenCalledTimes(1);
  });

  test('should show modal after prompt timeout', async () => {
    render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    // advance timers by 12 minutes
    jest.advanceTimersByTime(60 * 12 * 1000);

    expect(dialogOpenMock).not.toHaveBeenCalled();

    act(() => {
      // advance timers by 1 minute
      jest.advanceTimersByTime(60 * 1 * 1000);
      fireEvent.focus(document);
    });

    expect(dialogOpenMock).toHaveBeenCalledTimes(1);
  });

  test('should update modal header to reflect seconds before logout', () => {
    const { getByTestId } = render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    act(() => {
      // advance timers by 14 minutes
      jest.advanceTimersByTime(60 * 14 * 1000);
      fireEvent.focus(document);
    });

    const events = [
      {
        timeRemaining: 25,
        advanceTimersBySeconds: 35,
      },
      {
        timeRemaining: 20,
        advanceTimersBySeconds: 5,
      },
      {
        timeRemaining: 10,
        advanceTimersBySeconds: 10,
      },
      {
        timeRemaining: 5,
        advanceTimersBySeconds: 5,
      },
    ];

    for (const { timeRemaining, advanceTimersBySeconds } of events) {
      act(() => {
        // advance timers by timeRemaining
        jest.advanceTimersByTime(advanceTimersBySeconds * 1000);
        fireEvent.focus(document);
      });

      expect(getByTestId('modal-header').innerHTML).toEqual(
        `For security reasons, you'll be signed out in ${timeRemaining} seconds.`
      );
    }
  });

  test('should close modal when user clicks "Stay signed in"', () => {
    const { getByTestId } = render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    act(() => {
      // advance timers by 13 minutes
      jest.advanceTimersByTime(60 * 13 * 1000);
      fireEvent.focus(document);
    });

    fireEvent.click(getByTestId('modal-sign-in'));

    expect(dialogCloseMock).toHaveBeenCalledTimes(2);
  });

  test('should redirect when user clicks "Sign out"', async () => {
    const { getByTestId } = render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    act(() => {
      // advance timers by 13 minutes
      jest.advanceTimersByTime(60 * 13 * 1000);
      fireEvent.focus(document);
    });

    const clicked = fireEvent.click(getByTestId('modal-sign-out'));

    expect(clicked).toBe(true);
  });

  test('should reset remaining time to initial value when user clicks "stay signed in"', () => {
    const { getByTestId } = render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    act(() => {
      // advance timers by 14 minutes
      jest.advanceTimersByTime(60 * 14 * 1000);
      fireEvent.focus(document);
    });

    act(() => {
      // advance timers by 32 seconds
      jest.advanceTimersByTime(35_000);
      fireEvent.focus(document);
    });

    expect(getByTestId('modal-header').innerHTML).toEqual(
      `For security reasons, you'll be signed out in 25 seconds.`
    );

    const clicked = fireEvent.click(getByTestId('modal-sign-in'));

    expect(clicked).toBe(true);

    act(() => {
      // advance timers by 13 minutes
      jest.advanceTimersByTime(60 * 13 * 1000);
      fireEvent.focus(document);
    });

    expect(getByTestId('modal-header').innerHTML).toEqual(
      `For security reasons, you'll be signed out in 2 minutes.`
    );
  });

  test('should redirect when user does not interact after specified time', () => {
    render(
      <LogoutWarningModal
        promptBeforeLogoutSeconds={120} // 2 minutes
        logoutInSeconds={900} // 15 minutes
      />
    );

    act(() => {
      // set system time to 15 minutes into the future
      jest.setSystemTime(Date.now() + 60 * 15 * 1000);

      // advance timers by 15 minutes
      jest.advanceTimersByTime(60 * 15 * 1000);

      fireEvent.focus(document);
    });

    expect(routerPushMock).toHaveBeenCalledWith(
      '/auth/inactive?redirect=%2Ftemplates%2Ftesting'
    );

    expect(dialogCloseMock).toHaveBeenCalledTimes(1);
  });
});
