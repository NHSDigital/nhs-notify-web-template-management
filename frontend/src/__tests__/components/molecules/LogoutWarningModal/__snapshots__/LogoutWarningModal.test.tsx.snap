// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoutWarningModal should match snapshot 1`] = `
<DocumentFragment>
  <dialog
    aria-labelledby="idle-warning-heading"
    aria-live="assertive"
    aria-modal="true"
    class="modal"
    role="alertdialog"
  >
    <div
      class="modal__content"
    >
      <div
        class="modal__heading"
        data-testid="modal-header"
        id="idle-warning-heading"
      >
        For security reasons, you'll be signed out in 2 minutes.
      </div>
      <div
        class="modal__body"
      >
        <p>
          If you're signed out, any unsaved changes will be lost.
        </p>
      </div>
      <div
        class="modal__footer"
      >
        <button
          aria-disabled="false"
          class="nhsuk-button signIn"
          data-testid="modal-sign-in"
          type="submit"
        >
          Stay signed in
        </button>
        <div
          class="signOut"
        >
          <a
            class="nhsuk-link"
            data-testid="modal-sign-out"
            href="/auth/signout?redirect=%2Ftemplates%2Fcreate-and-submit-templates"
          >
            Sign out
          </a>
        </div>
      </div>
    </div>
  </dialog>
</DocumentFragment>
`;
